In [47]:
#import dataset
import keras
from keras.datasets import reuters

In [48]:
# split dataset
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=None, test_split=0.2)
word_index = reuters.get_word_index(path="reuters_word_index.json")

print('# of Training Samples: {}'.format(len(x_train)))
print('# of Test Samples: {}'.format(len(x_test)))

num_classes = max(y_train) + 1
print('# of Classes: {}'.format(num_classes))

# of Training Samples: 8982
# of Test Samples: 2246
# of Classes: 46


In [49]:
#map index to words
index_to_word = {}
for key, value in word_index.items():
    index_to_word[value] = key

In [50]:
print(' '.join([index_to_word[x] for x in x_train[0]]))
print(y_train[0])

the wattie nondiscriminatory mln loss for plc said at only ended said commonwealth could 1 traders now april 0 a after said from 1985 and from foreign 000 april 0 prices its account year a but in this mln home an states earlier and rise and revs vs 000 its 16 vs 000 a but 3 psbr oils several and shareholders and dividend vs 000 its all 4 vs 000 1 mln agreed largely april 0 are 2 states will billion total and against 000 pct dlrs
3


Next, we need to binarize our data

In [53]:
from keras.preprocessing.text import Tokenizer

max_words = 100

tokenizer = Tokenizer(num_words=max_words)
x_train = tokenizer.sequences_to_matrix(x_train, mode='binary')
x_test = tokenizer.sequences_to_matrix(x_test, mode='binary')

y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

In [54]:
print(x_train[0])
print(len(x_train[0]))

print(y_train[0])
print(len(y_train[0]))

[0. 1. 0. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 0. 1. 1. 1. 0. 1. 0. 0. 1. 0.
 0. 1. 1. 0. 0. 1. 1. 0. 1. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 1. 1. 0. 0. 0.
 1. 1. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 1. 0. 0. 0. 0. 1. 1. 0. 0. 0. 0. 0.
 0. 0. 0. 0.]
100
[0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
46


In [55]:
#check out the types of train data
print(x_train.shape)
print(y_train.shape)

(8982, 100)
(8982, 46)


In [56]:
#construct model
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
embedding_vecor_length = 100
max_review_length =100
num_of_words=100
model = Sequential()
model.add(keras.layers.Embedding(num_of_words, embedding_vecor_length, input_length=max_review_length))
model.add(keras.layers.LSTM(64,activation="tanh")) #with one LSTM layer
model.add(keras.layers.Dense(46, activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=3, batch_size=3)


Epoch 1/3
2994/2994 [==============================] - 108s 36ms/step - loss: 2.4208 - accuracy: 0.3499 - val_loss: 2.3611 - val_accuracy: 0.3620
Epoch 2/3
2994/2994 [==============================] - 106s 35ms/step - loss: 2.2542 - accuracy: 0.4127 - val_loss: 2.1434 - val_accuracy: 0.4844
Epoch 3/3
2994/2994 [==============================] - 107s 36ms/step - loss: 2.0884 - accuracy: 0.4773 - val_loss: 2.0597 - val_accuracy: 0.4907


In [58]:
score = model.evaluate(x_test, y_test, batch_size=batch_size, verbose=1)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

71/71 [==============================] - 1s 15ms/step - loss: 2.0597 - accuracy: 0.4907
Test loss: 2.0597403049468994
Test accuracy: 0.4906500577926636


In [59]:
#construct the model
model = Sequential()
model.add(keras.layers.Embedding(num_of_words, embedding_vecor_length, input_length=max_review_length))
model.add(keras.layers.LSTM(64,activation="tanh",return_sequences=True)) #with three LSTM layer
model.add(keras.layers.LSTM(64,activation="tanh",return_sequences=True))
model.add(keras.layers.LSTM(64,activation="tanh"))
model.add(keras.layers.Dense(46, activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=3, batch_size=3)

Epoch 1/3
2994/2994 [==============================] - 292s 98ms/step - loss: 2.4341 - accuracy: 0.3496 - val_loss: 2.4091 - val_accuracy: 0.3620
Epoch 2/3
2994/2994 [==============================] - 296s 99ms/step - loss: 2.3236 - accuracy: 0.3975 - val_loss: 2.3089 - val_accuracy: 0.4270
Epoch 3/3
2994/2994 [==============================] - 296s 99ms/step - loss: 2.1503 - accuracy: 0.4554 - val_loss: 2.0536 - val_accuracy: 0.4782


In [60]:
score = model.evaluate(x_test, y_test, batch_size=batch_size, verbose=1)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

71/71 [==============================] - 3s 41ms/step - loss: 2.0536 - accuracy: 0.4782
Test loss: 2.053628921508789
Test accuracy: 0.47818344831466675


In [37]:
model = Sequential()
model.add(keras.layers.Embedding(num_of_words, embedding_vecor_length, input_length=max_review_length))
model.add(keras.layers.LSTM(64,activation="tanh")) #with one LSTM layer and one dropout layer
model.add(keras.layers.Dropout(0.5))
model.add(keras.layers.Dense(46, activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=3, batch_size=3)

Epoch 1/3
2994/2994 [==============================] - 108s 36ms/step - loss: 2.4768 - accuracy: 0.3422 - val_loss: 2.3824 - val_accuracy: 0.3620
Epoch 2/3
2994/2994 [==============================] - 109s 36ms/step - loss: 2.3960 - accuracy: 0.3614 - val_loss: 2.3417 - val_accuracy: 0.3967
Epoch 3/3
2994/2994 [==============================] - 109s 36ms/step - loss: 2.3241 - accuracy: 0.4088 - val_loss: 2.2134 - val_accuracy: 0.4430


In [38]:
score = model.evaluate(x_test, y_test, batch_size=batch_size, verbose=1)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

71/71 [==============================] - 1s 15ms/step - loss: 2.2134 - accuracy: 0.4430
Test loss: 2.213419198989868
Test accuracy: 0.44300979375839233


In [61]:
model = Sequential()
model.add(keras.layers.Embedding(num_of_words, embedding_vecor_length, input_length=max_review_length))
model.add(keras.layers.LSTM(128,activation="tanh")) #with one LSTM layer
model.add(keras.layers.Dense(46, activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=3, batch_size=3)


Epoch 1/3
2994/2994 [==============================] - 194s 65ms/step - loss: 2.4189 - accuracy: 0.3510 - val_loss: 2.3546 - val_accuracy: 0.3856
Epoch 2/3
2994/2994 [==============================] - 192s 64ms/step - loss: 2.3205 - accuracy: 0.3929 - val_loss: 2.2642 - val_accuracy: 0.4359
Epoch 3/3
2994/2994 [==============================] - 192s 64ms/step - loss: 2.1987 - accuracy: 0.4424 - val_loss: 2.1598 - val_accuracy: 0.4519


In [62]:
score = model.evaluate(x_test, y_test, batch_size=batch_size, verbose=1)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

71/71 [==============================] - 2s 33ms/step - loss: 2.1598 - accuracy: 0.4519
Test loss: 2.1598076820373535
Test accuracy: 0.451914519071579


### Summary
Using one layer LSTM with 64 units has accuracy to be 0.49

The first atttemp I tried is to increase the number of LSTM layer, however, the accuracy decreased. 

The second attemp I made is to use dropout layer with single LSTM layer, however, the accuracy decreased. 

The last attemp I made is using sing LSTM layer with 128 units (double units comprared to the original model), however, the accuracy decreased. 

Overall, the simplest model with fewer units give the best accuracy. RNN, or other neural network is black box algorithm, adjusting the hyperparameters are tricky and challenging. 